### ISSUE TITLE
Support for Software MIDI synthesizers

### ISSUE THREAD

**Issue by putridpete** [InsanityBringer/ChocolateDescent#8]
Is it possible to make Chocolate Descent work with programs such as OmniMIDI and VirtualMIDISynth? I know I can edit chocolatedescent.cfg and point towards whatever soundfont I want to use, but I prefer more control over the whole process. Besides, I find fluidsynth to be a bit quiet, even when the volume is set at max in the game settings, at least when compared to the SC-55 OGG music pack I set up for the port.

**Comment by InsanityBringer** [comment]
It's possible, and I've been eyeballing it. My main reason initially for not adding support for MMEAPI midi synthesis is due to the poor user experience when using the default Windows synth (such as the volume problem that came into existence on Vista) and the hacks required to work around them, but I've gotten requests to be able to use hardware MIDI drivers, as well as these software synth drivers. 

I don't want to support it as the primary output method since I'm not too happy about having to write a midiproc style hack to enable basic MIDI playback, but I'm fine with implementing it to enable the use of these drivers.

**Comment by putridpete** [comment]
Thanks for your response. Of course I wouldn't expect for it to be the primary output method, since most source ports require tweaking anyway to be able to use that kind of software. I think Chocolate Doom, Crispy Doom and even EDuke32 use OPL3 emulation as their default method, but I've configured them to use OmniMIDI instead. If something similar was possible for Chocolate Descent it would really enhance the experience.

I don't know if it's possible to play custom missions in Chocolate Descent (sorry, I'm not a very technical person; my knowledge never extends past downloading source ports, setting them up and rudimentary edits on files using notepad/notepad++), but I would imagine this can help custom maps that feature their own MIDIs.

**Comment by InsanityBringer** [comment]
Well, I'm sorry it took over 2 years to get started on this, but an initial implementation has been started in commit [42752a8](https://github.com/InsanityBringer/ChocolateDescent/commit/42752a80f5a6d6ffb83a964df930262cc9db0739). It still needs a lot of ironing out, and it will only default to the Midi Mapper device. I'm not fully sure how to implement the UI on this one to select devices, if I'm being honest... I'll figure something out.

**Comment by putridpete** [comment]
Funny enough, in the time it took for this to get implemented I switched to Linux, learned stuff beyond rudimentary edits (like how to compile from source and the like) and I'm currently running Arch in my desktop and Gentoo in my laptop.

I also recently got rid of Windows altogether so I no longer have access to those synths. I'm currently using fluidsynth with the qsynth graphical front end, but I'm guessing that would be somewhat redundant. Either way, thanks a lot for implementing this. It's a huge feature to have. Considering this feature is now technically implemented, should this be closed or kept open until it's more complete?
